rollup/types/encoding/da.go

```diff
@@ -0,0 +1,290 @@
+func (c *Chunk) NumL2Transactions() uint64 {
+	var totalTxNum uint64
+	for _, block := range c.Blocks {
+		totalTxNum += block.NumL2Transactions()
+	}
+	return totalTxNum
+}
+
+// L2GasUsed calculates the total gas of L2 transactions in a Chunk.
+func (c *Chunk) L2GasUsed() uint64 {
+	var totalTxNum uint64
+	for _, block := range c.Blocks {
+		totalTxNum += block.Header.GasUsed
+	}
+	return totalTxNum
+}
+
+// StateRoot gets the state root after committing/finalizing the batch.
+func (b *Batch) StateRoot() common.Hash {
+	numChunks := len(b.Chunks)
+	if len(b.Chunks) == 0 {
+		return common.Hash{}
+	}
+	lastChunkBlockNum := len(b.Chunks[numChunks-1].Blocks)
+	return b.Chunks[len(b.Chunks)-1].Blocks[lastChunkBlockNum-1].Header.Root
+}
+
+// WithdrawRoot gets the withdraw root after committing/finalizing the batch.
+func (b *Batch) WithdrawRoot() common.Hash {
+	numChunks := len(b.Chunks)
+	if len(b.Chunks) == 0 {
+		return common.Hash{}
+	}
+	lastChunkBlockNum := len(b.Chunks[numChunks-1].Blocks)
+	return b.Chunks[len(b.Chunks)-1].Blocks[lastChunkBlockNum-1].WithdrawRoot
+}
+
+// TxsToTxsData converts transactions to a TransactionData array.
+func TxsToTxsData(txs types.Transactions) []*types.TransactionData {
+	txsData := make([]*types.TransactionData, len(txs))
+	for i, tx := range txs {
+		v, r, s := tx.RawSignatureValues()
+
+		nonce := tx.Nonce()
+
+		// We need QueueIndex in `NewBatchHeader`. However, `TransactionData`
+		// does not have this field. Since `L1MessageTx` do not have a nonce,
+		// we reuse this field for storing the queue index.
+		if msg := tx.AsL1MessageTx(); msg != nil {
+			nonce = msg.QueueIndex
+		}
+
+		txsData[i] = &types.TransactionData{
+			Type:       tx.Type(),
+			TxHash:     tx.Hash().String(),
+			Nonce:      nonce,
+			ChainId:    (*hexutil.Big)(tx.ChainId()),
+			Gas:        tx.Gas(),
+			GasPrice:   (*hexutil.Big)(tx.GasPrice()),
+			GasTipCap:  (*hexutil.Big)(tx.GasTipCap()),
+			GasFeeCap:  (*hexutil.Big)(tx.GasFeeCap()),
+			To:         tx.To(),
+			Value:      (*hexutil.Big)(tx.Value()),
+			Data:       hexutil.Encode(tx.Data()),
+			IsCreate:   tx.To() == nil,
+			AccessList: tx.AccessList(),
+			V:          (*hexutil.Big)(v),
+			R:          (*hexutil.Big)(r),
+			S:          (*hexutil.Big)(s),
+		}
+	}
+	return txsData
+}
+
+// DecodeBlockContext decodes the encoded block context byte array into BlockContext.
+func DecodeBlockContext(encodedBlockContext []byte) (*BlockContext, error) {
+	if len(encodedBlockContext) != 60 {
+		return nil, errors.New("block encoding is not 60 bytes long")
+	}
+
+	return &BlockContext{
+		BlockNumber:     binary.BigEndian.Uint64(encodedBlockContext[0:8]),
+		Timestamp:       binary.BigEndian.Uint64(encodedBlockContext[8:16]),
+		GasLimit:        binary.BigEndian.Uint64(encodedBlockContext[48:56]),
+		NumTransactions: binary.BigEndian.Uint16(encodedBlockContext[56:58]),
+		NumL1Messages:   binary.BigEndian.Uint16(encodedBlockContext[58:60]),
+	}, nil
```

Comment:
Seems like some fields are missing?

Reply:
added the `basefee` field and removed other non-encoded fields in [ef09494](https://github.com/scroll-tech/go-ethereum/pull/678/commits/ef09494ea638852fafbd1e247780ef7846e7219d).

Reply:
moved BlockContext into codecv0 and codecv1 because I found that the structure is the same as DABlock. related commit: [62509ed](https://github.com/scroll-tech/go-ethereum/pull/678/commits/62509edb1c6f51f475b30e8665b1b7988cf266b5).